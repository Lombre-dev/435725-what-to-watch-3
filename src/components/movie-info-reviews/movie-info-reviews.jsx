import PropTypes from 'prop-types';
import React from 'react';
import {connect} from 'react-redux';
import {setReviews} from '../../redux/reviews/actions';
import {Operations} from '../../redux/reviews/operations';
import {getMovieReviews} from '../../redux/reviews/selectors';
import {getRatingScore} from '../../utils/movie-utils';
import {Movie, Review} from '../types';

class MovieInfoReviews extends React.PureComponent {

  componentDidMount() {

    const {getReviews, movie} = this.props;

    getReviews(movie.id);
  }

  componentDidUpdate(prevProps) {

    const {movie: prevMovie} = prevProps;
    const {movie, getReviews} = this.props;

    if (movie !== prevMovie) {
      getReviews(movie.id);
    }
  }

  render() {

    const {reviews} = this.props;

    if (Boolean(reviews) === false) {
      return <></>;
    }

    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">
          {
            reviews.map((review) => {

              const date = new Date(review.date);

              return (
                <div key={review.date} className="review">
                  <blockquote className="review__quote">
                    <p className="review__text">{review.text}</p>

                    <footer className="review__details">
                      <cite className="review__author">{review.author}</cite>
                      <time className="review__date" dateTime={date.toISOString()}>{
                        date.toLocaleDateString(`en-US`, {month: `long`, day: `numeric`, year: `numeric`})
                      }</time>
                    </footer>
                  </blockquote>

                  <div className="review__rating">{getRatingScore(review.score)}</div>
                </div>
              );
            })
          }
        </div>
      </div>
    );
  }
}

MovieInfoReviews.propTypes = {
  movie: Movie.isRequired,
  reviews: PropTypes.arrayOf(Review),
  getReviews: PropTypes.func,
};

function mapStateToProps(state) {
  return {
    reviews: getMovieReviews(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    getReviews: (movieId) => {
      dispatch(setReviews([]));
      dispatch(Operations.getReviews(movieId));
    },
  };
}

export {MovieInfoReviews};
export default connect(mapStateToProps, mapDispatchToProps)(MovieInfoReviews);
