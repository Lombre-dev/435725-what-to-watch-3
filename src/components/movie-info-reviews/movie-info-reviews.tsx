import * as React from 'react';
import {connect} from 'react-redux';
import {Operations} from '../../redux/movie/operations';
import {getDetailedMovieReviews, getDetailedMovieStatus} from '../../redux/movie/selectors';
import {getRatingScore} from '../../utils/movie-utils';
import LoadingDataBlock from '../loading-data-block/loading-data-block';
import {TMovie, TReview, LoadingDataStatus} from '../../types';

type TProps = {
  status?: LoadingDataStatus;
  movie: TMovie;
  reviews?: TReview[];
  getReviews: (movieId: number) => void;
};

class MovieInfoReviews extends React.PureComponent<TProps> {

  public componentDidMount() {

    const {getReviews, movie} = this.props;

    getReviews(movie.id);
  }

  public componentDidUpdate(prevProps) {

    const {movie: prevMovie} = prevProps;
    const {movie, getReviews} = this.props;

    if (movie.id !== prevMovie.id) {
      getReviews(movie.id);
    }
  }

  public render() {

    const {status, reviews} = this.props;

    if (status !== LoadingDataStatus.READY) {
      return <LoadingDataBlock status={status} />;
    }

    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">
          {
            reviews.map((review) => {

              const date = new Date(review.date);

              return (
                <div key={review.date} className="review">
                  <blockquote className="review__quote">
                    <p className="review__text">{review.text}</p>

                    <footer className="review__details">
                      <cite className="review__author">{review.author}</cite>
                      <time className="review__date" dateTime={date.toISOString()}>{
                        date.toLocaleDateString(`en-US`, {month: `long`, day: `numeric`, year: `numeric`})
                      }</time>
                    </footer>
                  </blockquote>
                  <div className="review__rating">{getRatingScore(review.score)}</div>
                </div>
              );
            })
          }
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    status: getDetailedMovieStatus(state),
    reviews: getDetailedMovieReviews(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    getReviews: (movieId: number) => {
      dispatch(Operations.getReviews(movieId));
    },
  };
}

export {MovieInfoReviews};
export default connect(mapStateToProps, mapDispatchToProps)(MovieInfoReviews);
