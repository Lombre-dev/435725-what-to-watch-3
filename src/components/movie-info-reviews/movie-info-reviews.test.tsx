import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {MovieInfoReviews} from './movie-info-reviews';

const MOVIE = {
  id: 0,
  title: `The Grand Budapest Hotel`,
  genres: [`Drama`],
  year: 2014,
  poster: `/img/the-grand-budapest-hotel-poster.jpg`,
  frames: [`/img/the-grand-budapest-hotel-poster.jpg`],
  ratingScore: 8.9,
  ratingReviewsCount: 240,
  preview: `/samples/sintel_trailer-480p.mp4`,
  src: `/samples/sintel_trailer-480p.mp4`,
  description: `Description`,
  story: `Story`,
  director: `Wes Andreson`,
  duration: 99,
  actors: [
    `Bill Murray`,
    `Edward Norton`,
    `Jude Law`,
    `Willem Dafoe`,
    `Some Actor 1`,
    `Some Actor 2`,
  ],
};
const HANDLE_EVENT = () => {
  // eslint-disable-next-line no-console
  console.log(`handleTestEvent`);
};

describe(`<MovieInfoReviews />`, () => {

  it(`render should be match markup`, () => {

    const result = renderer
      .create(<MovieInfoReviews
        movie={MOVIE}
        getReviews={HANDLE_EVENT}
      />)
      .toJSON();

    expect(result).toMatchSnapshot();
  });
});
