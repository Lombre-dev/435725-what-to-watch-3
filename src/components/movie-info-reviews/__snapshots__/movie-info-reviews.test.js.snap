// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieInfoReviews /> render should be match markup 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Awesome text ...
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Some Reviewer
          </cite>
          <time
            className="review__date"
            dateTime="2020-02-25T00:22:20.667Z"
          >
            February 25, 2020
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        8,2
      </div>
    </div>
  </div>
</div>
`;
